src/headset/oculus.c

```diff
@@ -0,0 +1,531 @@
+  if (OVR_FAILURE(result))
+    return;
+
+  result = ovr_Create(&state.session, &state.luid);
+  if (OVR_FAILURE(result)) {
+    ovr_Shutdown();
+    return;
+  }
+
+  state.lastButtonState = 0;
+  state.isInitialized = true;
+  state.isMirrored = true;
+  state.mirrorTexture = NULL;
+  int i;
+  for (i = 0; i < 2; i++) {
+    state.eyeTextures[i].size.w = 0;
+    state.eyeTextures[i].size.h = 0;
+    state.eyeTextures[i].fboId = 0;
+    state.eyeTextures[i].depthId = 0;
+    state.eyeTextures[i].chain = NULL;
+  }
+  state.clipNear = 0.1f;
+  state.clipFar = 30.f;
+
+  vec_init(&state.controllers);
+
+  for (i = 0; i < 2; i++) {
+    Controller *controller = lovrAlloc(sizeof(Controller), lovrControllerDestroy);
+    controller->id = ovrHand_Left+i;
+    vec_push(&state.controllers, controller);
+  }
+
+  lovrEventAddPump(lovrHeadsetPoll);
+  atexit(lovrHeadsetDestroy);
+}
+
+void lovrHeadsetDestroy() {
+  int i;
+  Controller *controller;
+  vec_foreach(&state.controllers, controller, i) {
+    lovrRelease(&controller->ref);
+  }
+
+  vec_deinit(&state.controllers);
+
+  if (state.mirrorTexture) {
+    ovr_DestroyMirrorTexture(state.session, state.mirrorTexture);
+    state.mirrorTexture = NULL;
+  }
+
+  for (i = 0; i < 2; i++) {
+    if (state.eyeTextures[i].chain) {
+      ovr_DestroyTextureSwapChain(state.session, state.eyeTextures[i].chain);
+      state.eyeTextures[i].chain = NULL;
+    }
+  }
+
+  ovr_Destroy(state.session);
+  ovr_Shutdown();
+
+  state.isInitialized = false;
+  state.frameIndex = 0;
+}
+
+static void checkInput(Controller *controller, int diff, int state, ovrButton button, ControllerButton target) {
+  if ((diff & button) > 0) {
+    Event e;
+    if ((state & button) > 0) {
+      e.type = EVENT_CONTROLLER_PRESSED;
+      e.data.controllerpressed.controller = controller;
+      e.data.controllerpressed.button = target;
+    }
+    else {
+      e.type = EVENT_CONTROLLER_RELEASED;
+      e.data.controllerreleased.controller = controller;
+      e.data.controllerreleased.button = target;
+    }
+    lovrEventPush(e);
+  }
+}
+
+void lovrHeadsetPoll() {
+  ovr_SetTrackingOriginType(state.session, ovrTrackingOrigin_FloorLevel);
+  state.ts = ovr_GetTrackingState(state.session, 0, true);
+  ovr_GetInputState(state.session, ovrControllerType_Touch, &state.is);
+
+  ovrSessionStatus status;
+  ovr_GetSessionStatus(state.session, &status);
+
+  if (status.ShouldRecenter) {
+    ovr_RecenterTrackingOrigin(state.session);
+    ovr_ClearShouldRecenterFlag(state.session);
+  }
+
+  if (status.ShouldQuit) {
+    Event e;
+    e.type = EVENT_QUIT;
+    e.data.quit.exitCode = 0;
+    lovrEventPush(e);
+  }
+
+  //if (!status.HmdMounted) // update mirror only?
+  //ovr_SetInt(state.session, OVR_PERF_HUD_MODE, (int)ovrPerfHud_PerfSummary);
+
+  state.hmdPresent = (status.HmdPresent == ovrTrue)? true : false;
+
+  Controller* left = state.controllers.data[ovrHand_Left];
+  Controller* right = state.controllers.data[ovrHand_Right];
```

Comment:
I think this is only guaranteed to work if `ovrHand_Left` is 0 and `ovrHand_Right` is 1.

Reply:
ovrHand_Left and _Right are defined per the docs to be used as 0/1 for array access